Namari/interest feature

amari-n opened the pull request:
**Changes Made in this PR:**

I have added an interest rate feature to the Testudo Bank which will apply interest to a user's balance after 5 deposits greater that $20.00. The purpose of this change is to act as an incentive for potential customers to bank with Testudo Bank. The interest rate feature will reward the loyalty of customers and will help create an influx of new customers who want to use Testudo Bank.
<hr>

`src/main/java/net/testudobank/MvcController.java`
- Added an if statement to `submitDeposit()` that will check 3 restrictions: (1) if the user's deposit is larger than $20.00, (2) if the user has no overdraft balance, and (3) if the user has a positive balance. These are the three prerequisites required to increment a user's `NumDepositsForInterest` and apply interest.
- Implemented `applyInterest()` which will increment a user's `NumDepositsForInterest` column in the customers table, and will increase the user's cash balance by applying `BALANCE_INTEREST_RATE` only if `userNumberOfDepositsForInterest` is a multiple of 5.
<hr>

`src/test/java/net/testudobank/tests/MvcControllerIntegTest.java`
- Wrote integration test `testInterestAppliedAfter5thTransaction()` that will verify that interest is being applied to the user's cash balance after the 5th transaction of over $20.00.
<hr>

**Future Changes to Make:**

I will need to implement more integration tests to verify that:
- Interest is not being applied every transaction after the 5th
- Interest is only being applied to deposits over $20.00
- Interest is not being applied if the deposit is exactly $20.00

s1ddh4rthc commented:
Hey Nizar, nice job with this PR. I think an improvement you can make to your PR is to make the logic so that you reset the numDepositsForInterest to 0 after interest is applied, rather than checking if the number % 5 == 0. Same logic, but I think that the approach I mentioned looks cleaner. Make sure you write your integ tests for the second part.

amari-n commented:
**Changes made in my most recent commit:**
I addressed the specific comments you made, and further increased readability of my code. I have also added 6 more integration tests that test the functionality of my interest rate feature, as well as the basic edge cases mentioned in the design doc.

<hr>

`src/main/java/net/testudobank/MvcController.java`

- Made the static constant `NUM_DEPOSITS_REQUIRED_FOR_INTEREST` in order to increase readability
- Addressed your comment to make `numDepositsForInterest` reset to 0 after the user's 5th deposit of over $20
- Made other minor changes to make it easier to read/understand my code

<hr>

`src/test/java/net/testudobank/tests/MvcControllerIntegTest.java`

- Made the static constant `BALANCE_INTEREST_RATE` to increase readability
- Wrote integ test `testInterestNotAppliedForTransactionsAfter5th()` that verifies that interest is not applied to the user's balance for the 4 immediate deposits after the 5th deposit. Also checks that `numDepositsForInterest` resets to 0 after the 5th deposit.
- Wrote integ test `testInterestAppliedOnlyOnDepositsAbove20Dollars()` that verifies that interest is only being applied to the user's balance on deposits that are larger than $20 after the 5th deposit
- Wrote integ test `testInterestNotBeingAppliedIfDepositIsExactly20Dollars()` that verifies that interest is not being applied to the user's balance when deposits are made of exactly $20. Also verifies that deposits of exactly $20 are not incrementing `numDepositsForInterest`
- Wrote integ test `testOnePennyAbove20DollarsWillIncrementsUserNumDepositsForInterest()` that verifies the edge case that deposits of $20.01 will increment `numDepositsForInterest` and apply interest for the user's balance after the 5th deposit
- Wrote integ test `testOnePennyBelow20DollarsWillNotIncrementUserNumDepositsForInterest()` that verifies the edge case that deposits of $19.99 are not incrementing `numDepositsForInterest` and will not apply interest to the user's balance.
- Wrote integ test `testNoInterestAppliedToCustomersInOverdraftOrWithNoMoney()` that verifies that users who are in overdraft or who have no money are not being applied interest.
